pkg/http/server.go

```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
 	}
 	return ""
 }
+
+func (s *Server) existCertificate(uDomain string) bool {
+	if _, err := os.Stat("./lets_encrypt/" + uDomain + ".key"); os.IsNotExist(err) {
+		return false
+	}
+	if _, err := os.Stat("./lets_encrypt/" + uDomain + ".crt"); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func (s *Server) getLetsEncrypt(uDomain string) error {
+	letsEncryptURL := utils.LetsEncryptStaging
+	certDirectory := "./lets_encrypt"
+	if s.Production {
+		letsEncryptURL = utils.LetsEncryptProduction
+	}
+	err := os.MkdirAll(certDirectory, 0644)
+	if err != nil {
+		return err
+	}
+	// Generate account key
+	aKey, err := newKey(path.Join(certDirectory, "account.pem"))
+	if err != nil {
+		return err
+	}
+
+	// Create the ACME client
+	c := &acme.Client{
+		Key:          aKey,
+		DirectoryURL: letsEncryptURL,
+	}
+
+	// Begin with registration
+	log.Infof("Attempting registration...")
+	_, err = c.Register(context.TODO(), nil, func(string) bool { return true })
+	if err != nil {
+		return err
+	}
+	log.Infof("Registration succeeded")
+
+	// Attempt authorization
+	auth, err := c.Authorize(context.TODO(), s.Domain)
+	if err != nil {
+		return err
+	}
+	var challenge *acme.Challenge
+	for _, c := range auth.Challenges {
+		if c.Type == "http-01" {
```

Comment:
http-01 means a chalange over unauthenticated HTTP and port 80, which means people who want to listen only on port 443 will have to open one more port in the firewall. let's talk more about this offline.